### ISSUE TITLE
int value will not accepted for any in type(def/set)/callback

### ISSUE THREAD

**Issue by Bara** [alliedmodders/sourcepawn#613]
int will no more accept for any in callbacks/type(def/set) in latest sourcepawn update on sm1.11?

as example with [ripext](https://github.com/ErikMinekus/sm-ripext)
[typeset](https://github.com/ErikMinekus/sm-ripext/blob/main/pawn/scripting/include/ripext/http.inc#L67-L71): `function void (HTTPResponse response, any value);`

```
request.Post(json, CallbackName, someInt);

public void CallbackName(HTTPResponse response, int someInt, const char[] error)
{
  [...]
}
```

throws the error: `error 100: function prototypes do not match`

by replacing `int someInt` with `any someInt` it'll fix this error.

Example for reproducing this error:
```
public void OnPluginStart()
{
    int iTest = 1;

    RequestFrame(FrameTest, iTest);
}

public void FrameTest(int test)
{
    
}
```

**Comment by dvander** [comment]
This is intentional. The callback needs to accept "any". But the callback you've given only accepts "int".

I'd like to keep this rule, but I'm honestly on the fence about it. The motivation is that we can't type check the implicit assignment, but (1) other uses of "any" lack these checks and (2) I want to add the dynamic checks at some point in the future.

So... I could be convinced to drop this.

**Comment by Bara** [comment]
I've nothing against this change, but wouldn't it break a lot of plugins when someone want to compile it against the latest changes?

And compiled plugins against SM1.10 ( or lower ) will still work?

**Comment by dvander** [comment]
We don't guarantee source compatibility. Not yet, at least. When making changes like this I check it against a large corpus of available plugins to see how widespread the breakage is. This one didn't break anything. The corpus is not 100% representative, so if there's additional breakage, we can consider reverting this.

**Comment by c0rp3n** [comment]
This new behaviour doesnt feel great for writing readable code, forcing the callback functions types to have to be any makes code more convoluted than it was before as you will also now need to include the type in the name of the parameter for some functions.

```sourcepawn
void Timer_Rewards(Handle timer, int amount);  // clear
void Timer_Rewards(Handle timer, any amount);  // is unclear (int or float?)
void Timer_Rewards(Handle timer, any iAmount); // is now clear a int should be passed
```

The latest diff for TTT shows the disadvantages quite well https://github.com/TroubleInTerroristTown/Public/commit/31443ec3a3f6b1e0210931fadd11a8e1ea0116a4.

**Comment by dvander** [comment]
This can be resolved for the moment by adding a new signature into the Timer typeset.

**Comment by DarthMan** [comment]
This change is very bad. There are many cases in which it would require to write additional code, for example if using TR_TraceRayFilterEx, the last parameter is of type any. If you pass the player or entity id, you would have to create an additional variable, e.g. any aEnt = view_as<any>(iEnt); . This would break many existing plugins and would also mean creating new useless variables.

**Comment by dvander** [comment]
Of the ~10,000 plugins in my test suite, this broke zero of them. If you have an example on the forums please let me know and I'll investigate.

**Comment by dysphie** [comment]
I'm not a fan of this change either. It's pretty common practice (at least for me) to detail the type in the callback prototype. I think it adds to the overall readability of the plugin. Here's a few public plugins affected by this:

- https://github.com/dysphie/nmrih-team-healing/blob/main/scripting/nmrih-team-healing.sp#L1040 
	- https://forums.alliedmods.net/showthread.php?p=2751622#post2751622
- https://github.com/dysphie/sm-map-translator/blob/f74e130aa23e08bf50d768fcce45ec068f0e16f2/scripting/map-translator.sp#L241
	- https://forums.alliedmods.net/showthread.php?t=333658
- https://github.com/dysphie/nmrih-voicecmd-block/blob/8b395e71dcd0112f473b7a7ac9940b762784b541/nmrih-voicecmd-block.sp#L75
	- https://forums.alliedmods.net/showthread.php?p=2752182#post2752182
- https://github.com/dysphie/sourcemod-indev/blob/9026fc4feb0b4d39fa1f639962221399ad77d9e5/incap-wip.sp#L628

**Comment by DarthMan** [comment]
For example I have this piece of code:

`Handle hTrace;

hTrace = TR_TraceRayFilterEx(fVecTraceStart, fVecTraceMaxEndPos, EntMod_GetBasicMaskType(), RayType_EndPoint, TraceRayDontHitSelf, iID);
TR_GetEndPosition(fVecAimEndPos, hTrace);

delete hTrace;`

This would no longer compile and throw an error, because iID is of type int rather than any. The iID variable is the player id.

**Comment by dvander** [comment]
Thanks. I'll start expanding the compatibility testing tool to include github repos.

The problem with these casts is that they break contravariance. The callback signature must accept `any`, but you're giving it a callback that only accepts a narrowed type. This would fail in almost any other statically typed language. Normally you'd be forced to add an explicit dynamic cast, and that was the original intent with SourcePawn. Otherwise, we would not have introduced the "typeset" or "funcenum" keywords which have been there since 1.0. It just so happened that signature type checking has basically never worked, and we didn't notice until 1.11.

On the other hand, folks here have presented a very strong and clear argument that it improves usability. Why force programmers to do the cast if the runtime could be implicitly enforcing them? Of course, the runtime can't do that yet. But I think it could in the future, so I'll restore the bidirectional cast to/from `any`.

**Comment by JoinedSenses** [comment]
this would break a bunch of my stuff as well. lots of folks these days arent hosting their files on AM, but repos like GH - meaning you wont get an accurate representation of what code patterns are being used if only searching the forums.

**Comment by dvander** [comment]
FYI you can now get your git repos included in the compatibility corpus. This gets run (by me) after every compiler change. Just submit a PR that changes this file:

https://github.com/alliedmodders/sourcepawn/blob/master/tools/corpus/manifest.yaml

Inclusion does *not* mean we'll guarantee source compatibility for every plugin - it just gives us another data point for how SourcePawn is being used. If we do have to break source compatibility, affected plugins get dropped from the corpus. This tends to only happen for very old, unmaintained plugins though.

**Comment by JoinedSenses** [comment]
will it work for repos that have external dependencies that arent included?

**Comment by dvander** [comment]
What's the external dependency?

**Comment by JoinedSenses** [comment]
Varies per plugin, usually noted in the readme where to find it. Example here with my largest project <https://github.com/JoinedSenses/TF2-ECJ-JumpAssist>, which is one of a few I'd consider submitting.

**Comment by dvander** [comment]
There's a section in the manifest to put global dependencies, as long as they don't introduce files that would conflict with other imports.

**Comment by Bara** [comment]
Conflicts like already included includes by the manifest file or will the manifest include files overridden by the plugin/project versions? If yes, [TTT ](https://github.com/TroubleInTerroristTown/Public) could be helpful for future testing as larger project

**Comment by dvander** [comment]
Right now the tool will check out all the projects in the manifest. If they're marked as includes, it will add them to the include path used to test *all* plugins. So if two projects are marked as include projects, and they introduce the same exact .inc name, that would be a conflict we can't handle.

If you have a large project that has a complex build procedure, that is also not handled. If you're unsure, file a request and I'll figure out how to deal with it.

**Comment by Bara** [comment]
> So if two projects are marked as include projects, and they introduce the same exact .inc name, that would be a conflict we can't handle.

Or is it possible to add a single include file instead of the whole include directory? In the [current version](https://github.com/alliedmodders/sourcepawn/blob/master/tools/corpus/manifest.yaml) it includes the whole directory and how works the [corpus_skip.list](https://github.com/alliedmodders/sourcepawn/blob/6b487bce370b4a16b38b05916a51ccda414bb430/tools/corpus/corpus_skip.list)?

**Comment by dvander** [comment]
It's not yet possible to specify a single include file. The current tool, plus 1-2 files manually downloaded off the forums, gets something like 98% of plugins working. Of the ~15,000 plugins in the current test base, only ~300 fail to compile due to missing includes, and many of those are one-offs that are hard to track down. I'll happily take PRs to improve the tool though.

corpus_skip.list is the list of files known to not compile, so the tool will skip trying them on the next run. This is how I test for compatibility breakage.

**Comment by dvander** [comment]
This has been fixed on 1.11-dev. It won't be immediately fixed on master since the development branch is undergoing a lot of refactoring.
